packages/foam-vscode/src/features/folder-rename-handler.ts

```diff
@@ -0,0 +1,1112 @@
+        );
+
+        for (const fileUri of newFilesToAdd) {
+          try {
+            await this.foam.workspace.fetchAndSet(fileUri);
+            Logger.debug(
+              `Pre-added new file URI to workspace: ${fileUri.path}`
+            );
+          } catch (addError) {
+            Logger.warn(
+              `Failed to pre-add new file URI ${fileUri.path} to workspace:`,
+              addError
+            );
+          }
+        }
+
+        // Refresh matcher and graph
+        await this.foam.services.matcher.refresh();
+        await this.foam.graph.update();
+
+        Logger.debug('Pre-refresh complete');
+      } catch (preRefreshError) {
+        Logger.warn(
+          'Pre-refresh failed, continuing with calculation:',
+          preRefreshError
+        );
+      }
+
+      let dryRunCalcResult: DryRunCalculationResult | undefined;
+      try {
+        dryRunCalcResult = await this.calculateLinkUpdateEdits(
+          oldUri,
+          newUri,
+          markdownFilesInFolder
+        );
+        Logger.debug(
+          `Dry run complete: ${dryRunCalcResult.linksUpdatedCount} potential links in ${dryRunCalcResult.filesToEditCount} files.`
+        );
+        if (dryRunCalcResult.calculationErrors.length > 0) {
+          Logger.warn(
+            'Dry run encountered errors:',
+            dryRunCalcResult.calculationErrors.join('\n')
+          );
+          result.warnings.push(
+            ...dryRunCalcResult.calculationErrors.map(e => `Dry run: ${e}`)
+          );
+        }
+      } catch (dryRunError) {
+        Logger.error('Critical error during dry run calculation:', dryRunError);
+        result.errors.push(`Critical dry run error: ${dryRunError}`);
+      }
+      if (mode === 'always' && !options.confirmAction) {
+        Logger.debug(
+          "Mode is 'always', proceeding with link updates automatically."
+        );
+        userDecision = { action: 'proceed' }; // Simulate proceed for automatic application
+      } else if (mode !== 'never') {
+        // This covers 'confirm' or ('always' with confirmAction)
+        // If mode is 'never', userDecision is already set to skip, and this block is skipped.
+        userDecision = await FolderRenameDialog.show(
+          markdownFilesInFolder.length,
+          oldUri,
+          newUri,
+          dryRunCalcResult,
+          mode // Pass the current mode
+        );
+        if (userDecision.action === 'settings') {
+          Logger.debug(
+            'User chose to open settings for folder rename behavior. Aborting the current rename operation.'
+          );
+          // Open VS Code settings and search for the foam folder rename setting
+          await vscode.commands.executeCommand(
+            'workbench.action.openSettings',
+            'foam.folderRename.updateLinks'
+          );
+
+          // Abort the current rename operation by reverting the folder
+          try {
+            await this.revertFolderRename(newUri, oldUri);
```

Comment:
why do you need to revert the folder rename? isn't this called _before_ the renaming happens? (`onWillRenameFiles`)